src/collision/narrow_phase.rs

```diff
@@ -279,16 +279,31 @@ fn generate_constraints<C: AnyCollider>(
         let body2_bundle = collider2
             .parent
             .and_then(|p| narrow_phase.body_query.get(p.get()).ok());
-        if let (Some(body1), Some(body2)) = (
-            body1_bundle.map(|(body, _)| body),
-            body2_bundle.map(|(body, _)| body),
+        if let (Some((body1, rb_collision_margin1)), Some((body2, rb_collision_margin2))) = (
+            body1_bundle.map(|(body, rb_collision_margin1, _)| (body, rb_collision_margin1)),
+            body2_bundle.map(|(body, rb_collision_margin2, _)| (body, rb_collision_margin2)),
         ) {
             // At least one of the bodies must be dynamic for contact constraints
             // to be generated.
             if !body1.rb.is_dynamic() && !body2.rb.is_dynamic() {
                 continue;
             }
 
+            // Use the collider's own collision margin if specified, and fall back to the body's
+            // collision margin.
+            //
+            // The collision margin adds artificial thickness to colliders for performance
+            // and stability. See the `CollisionMargin` documentation for more details.
+            let collision_margin1 = collider1.collision_margin.map_or(
+                rb_collision_margin1.map_or(0.0, |margin| margin.0),
+                |margin| margin.0,
+            );
+            let collision_margin2 = collider2.collision_margin.map_or(
+                rb_collision_margin2.map_or(0.0, |margin| margin.0),
+                |margin| margin.0,
+            );
```

Comment:
```suggestion
            let collision_margin1 = collider1.collision_margin
                .or(rb_collision_margin1)
                .unwrap_or_default().0;
            let collision_margin2 = collider2.collision_margin
                .or(rb_collision_margin2)
                .unwrap_or_default().0;
```

The `map_or` chains are very unreadable imo, so I'd rewrite them like this

Reply:
You can't `unwrap_or_default` on `&CollisionMargin` (a reference), so it would become this:
```rust
let collision_margin1 = collider1
    .collision_margin
    .or(rb_collision_margin1)
    .copied()
    .unwrap_or_default()
    .0;
let collision_margin2 = collider2
    .collision_margin
    .or(rb_collision_margin2)
    .copied()
    .unwrap_or_default()
    .0;
```
I don't think that's better

Reply:
You could swap the `.copied().unwrap_or_default()` for a `.map_or(0.0, |margin| margin.0)`, but I very much would appreciate leaving the `or` for readability. Maybe it's just me, but chained `map_or`s induce way more mental load for me than function chains.